# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/ContentChild.cpp
# Commit: 0a44c410b59b
# Full Hash: 0a44c410b59bbc54bfd80ce598562e4362961d7e
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-04-24 09:51:27
# Regressor Bug: 1580565
# File Overlap Count: 3
# Description:
#   Bug 1580565 - Part 4: Use WindowContext to manage BrowsingContext cached status, r=farre
#   
#   The existing infrastructure which stored cached BrowsingContexts on the
#   BrowsingContextGroup was added before WindowContexts were added, and can cause
#   racing issues with partially discarded trees during process switches.
# ==============================================================================

diff -r 20dbcfc9eacc -r 0a44c410b59b dom/ipc/ContentChild.cpp
--- a/dom/ipc/ContentChild.cpp	Thu Apr 23 21:52:44 2020 +0000
+++ b/dom/ipc/ContentChild.cpp	Thu Apr 23 21:52:47 2020 +0000
@@ -3675,75 +3675,48 @@
   return IPC_OK();
 }
 
-mozilla::ipc::IPCResult ContentChild::RecvCacheBrowsingContextChildren(
-    const MaybeDiscarded<BrowsingContext>& aContext) {
-  if (aContext.IsNullOrDiscarded()) {
-    return IPC_OK();
-  }
-
-  aContext.get()->CacheChildren(/* aFromIPC */ true);
-  return IPC_OK();
-}
-
-mozilla::ipc::IPCResult ContentChild::RecvRestoreBrowsingContextChildren(
-    const MaybeDiscarded<BrowsingContext>& aContext,
-    const nsTArray<MaybeDiscarded<BrowsingContext>>& aChildren) {
-  if (aContext.IsNullOrDiscarded()) {
-    return IPC_OK();
-  }
-
-  nsTArray<RefPtr<BrowsingContext>> children(aChildren.Length());
-  for (const auto& child : aChildren) {
-    if (!child.IsNullOrDiscarded()) {
-      children.AppendElement(child.get());
-    }
-  }
-
-  aContext.get()->RestoreChildren(std::move(children), /* aFromIPC */ true);
-  return IPC_OK();
-}
-
 mozilla::ipc::IPCResult ContentChild::RecvRegisterBrowsingContextGroup(
-    nsTArray<BrowsingContext::IPCInitializer>&& aInits,
-    nsTArray<WindowContext::IPCInitializer>&& aWindowInits) {
+    nsTArray<SyncedContextInitializer>&& aInits) {
   RefPtr<BrowsingContextGroup> group = new BrowsingContextGroup();
+
   // Each of the initializers in aInits is sorted in pre-order, so our parent
   // should always be available before the element itself.
-  for (auto& init : aInits) {
+  for (auto& initUnion : aInits) {
+    switch (initUnion.type()) {
+      case SyncedContextInitializer::TBrowsingContextInitializer: {
+        auto& init = initUnion.get_BrowsingContextInitializer();
 #ifdef DEBUG
-    RefPtr<BrowsingContext> existing = BrowsingContext::Get(init.mId);
-    MOZ_ASSERT(!existing, "BrowsingContext must not exist yet!");
-
-    RefPtr<BrowsingContext> parent = init.GetParent();
-    MOZ_ASSERT_IF(parent, parent->Group() == group);
+        RefPtr<BrowsingContext> existing = BrowsingContext::Get(init.mId);
+        MOZ_ASSERT(!existing, "BrowsingContext must not exist yet!");
+
+        RefPtr<WindowContext> parent = init.GetParent();
+        MOZ_ASSERT_IF(parent, parent->Group() == group);
 #endif
 
-    bool cached = init.mCached;
-    RefPtr<BrowsingContext> ctxt =
-        BrowsingContext::CreateFromIPC(std::move(init), group, nullptr);
-
-    // If the browsing context is cached don't attach it, but add it
-    // to the cache here as well
-    if (cached) {
-      ctxt->Group()->CacheContext(ctxt);
-    } else {
-      ctxt->Attach(/* aFromIPC */ true);
+        RefPtr<BrowsingContext> ctxt =
+            BrowsingContext::CreateFromIPC(std::move(init), group, nullptr);
+        ctxt->Attach(/* aFromIPC */ true);
+        break;
+      }
+      case SyncedContextInitializer::TWindowContextInitializer: {
+        auto& init = initUnion.get_WindowContextInitializer();
+#ifdef DEBUG
+        RefPtr<WindowContext> existing =
+            WindowContext::GetById(init.mInnerWindowId);
+        MOZ_ASSERT(!existing, "WindowContext must not exist yet!");
+        RefPtr<BrowsingContext> parent =
+            BrowsingContext::Get(init.mBrowsingContextId);
+        MOZ_ASSERT(parent && parent->Group() == group);
+#endif
+
+        WindowContext::CreateFromIPC(std::move(init));
+        break;
+      };
+      default:
+        MOZ_ASSERT_UNREACHABLE();
     }
   }
 
-  for (auto& init : aWindowInits) {
-#ifdef DEBUG
-    RefPtr<WindowContext> existing =
-        WindowContext::GetById(init.mInnerWindowId);
-    MOZ_ASSERT(!existing, "WindowContext must not exist yet!");
-    RefPtr<BrowsingContext> parent =
-        BrowsingContext::Get(init.mBrowsingContextId);
-    MOZ_ASSERT(parent && parent->Group() == group);
-#endif
-
-    WindowContext::CreateFromIPC(std::move(init));
-  }
-
   return IPC_OK();
 }
 